### PR TITLE
[CS2113T-T13-3] Hospital management system

### PR DESCRIPTION
**Pull request by JustinnT**
AY1920S2-CS2113T-T13-3 PR request 

Team members: @brandoncjh @yukilite @andy-aw-why @thanhduc2000 @JustinnT

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,34 +1,431 @@
...
+
+#### 2.2.1 SAM record module
+
+###### [Back to top](#table-of-content)
+
+#### 2.2.2 BRANDON storage module
+
+###### [Back to top](#table-of-content)
+
+#### 2.2.3 A&D command module
```
Comment (btricec) [inline docs/DeveloperGuide.md]: The diagram is very complicated, the text is very small and the color makes it hard to read.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,34 +1,431 @@
...
+creates the connection from the ```Main``` class to the other classes required such as ```Storage``` class,
+```PatientList``` class and the ```Appointment``` class to name a few. 
+
+##### 2.2.3.1 AddPatientClass
+
+To add a patient, the ```AddPatientCommand``` class is used. For this ```AddPatientCommand``` class, it serves as the 
+façade class for the ```Main```, ```Patient``` , ```PatientList``` and the ```Storage``` class to interact with one 
+another. 
+
+![](images/AddPatientDiagram.png)
```
Comment (btricec) [inline docs/DeveloperGuide.md:251]: This diagram is also very complicated. The overlapping arrows are hard to follow.

**Review by btricec — COMMENTED** [review]
I think the DG is very well done. It's very comprehensive and explains the implementation well.
However, I think some of the class diagrams are very hard to follow and not too necessary.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,34 +1,614 @@
...
+flow in logical executions as these methods can be called whenever necessary.
+
+##### 2.2.1.1 Process of Object Creation
+Due to the nature of the above classes containing only getter and setter methods, following how the components interact 
+with each other would provide more accuracy in understanding how these classes are called and the role of its 
+methods. 
+To illustrate, the below example is used:
+* editp \index 1 \name Justin \age 69 \Pasir Panjang
+
+![](images/SD_Patient.png)
```
Comment (synCKun) [inline docs/DeveloperGuide.md]: i think the return arrows should be dotted :o 
Also, for HAMS, the second activation bar i think should be called by an arrow pointing at the the head of the activation bar instead of from it. 
the input command `editp \index 1 \name Justin \age 69 \Pasir Panjang` is helpful for reader's understanding,maybe a note box instead of the arrow into HAMS would make your diagram better?
example:

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,34 +1,614 @@
...
+
+###### [Back to top](#table-of-content)
+
+#### 2.2.2 Converter Module
+The converter module consists of one class which converts the format of date and time using a custom format defined by
+special formatting characters (ie. SimpleDateFormat). This class is primarily used to format a user-input date and time
+in the `Appointment` class. As illustrated below, its methods are called during the creation of the `Appointment`
+object constructor. 
+
+![](images/SD_Converter.png)
```
Comment (synCKun) [inline docs/DeveloperGuide.md]: For the activation bar of :TimeConverter, i think when object returns, the activation bar ends at the root return arrow.
What is the first Appointment object return arrow returning? 🤔

**Review by synCKun — COMMENTED** [review]
DG Review

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (JTWang2000) [inline docs/DeveloperGuide.md]: image

Maybe this part should be consecutive ? The activation bar should remain unbroken from the point the method is called until the method returns.

Comment (JTWang2000) [inline docs/DeveloperGuide.md]: image

Maybe you can decrease the length of the activation bar? 
 [Common notation error] Activation bar too long: The activation bar of a method cannot start before the method call arrives and a method cannot remain active after the method had returned. (From week 10 Topic 1)

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,34 +1,614 @@
...
+
+The Storage module consists of 3 different classes. 
+The PatientList and AppointmentList classes act as data structures to store the records of Patient and Appointment 
+objects respectively. They function as ADTs, where various commands from Command objects can manipulate the records within.
+The Storage class manages the load and save operations involving the PatientList and PatientList class. 
+These operations are usually invoked on startup, whenever changes are made to the ADTs and before exiting the program.
+The class diagram for the storage module is as seen below: 
+
+
+![](images/storageclass.PNG)
```
Comment (JTWang2000) [inline docs/DeveloperGuide.md:185]: image
Maybe you can consider to make the activation bar be consecutive? The activation bar should remain unbroken from the point the method is called until the method returns.

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,34 +1,614 @@
...
+5. When the above operation is successful, it will call upon the ```Ui``` class’ 
+```showPatientAddSuccess()``` function to display the success of adding the ```Patient``` object into the patient list.
+
+If the supplied patient age is a word or is missing, the age will be set to ```-1```. This value is chosen to indicate 
+that there isn’t a valid age set. Thus, when displaying the age, if ```-1``` is encountered, show age as an empty string 
+instead.
+
+Below shows the sequence diagram for ```AddPatientCommand``` class
+
+![](images/AddPatientCommandSequence.png)
```
Comment (JTWang2000) [inline docs/DeveloperGuide.md:278]: image
When geting the record, do you mean calling the method `getPatientRecord` of the patient class? If you are calling the method from the storage class, maybe the arrow should points to the storage.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md]: image
Consider ending the activation bar for `:EditPatientCommand` at where the return arrow starts?
Also consider adding/omitting return arrows consistently for the whole diagram? For example, there is no return arrow for the last activation bar of `:Storage` but there is a return arrow for the activation bar of `:Patient`

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md]: image

Consider ending the activation bar for `:TimeConverter` and `:Appointment` at where the return arrows start?
Also, should a return arrow be added for `:AddAppointmentCommand`?

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md:185]: image

Consider making the font size of `patientList` and `appointmentList` consistent?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,34 +1,614 @@
...
+
+Sequence Diagram for error checking when `DukeExpcetion` is called
+
+![](images/SD_parser/Slide10.JPG)
+
+Sequence Diagram for calling an enum
+
+![](images/SD_parser/Slide11.JPG)
+
+Sequence Diagram for error checking when `DukeExpcetion` is called
```
Comment (JTWang2000) [inline docs/DeveloperGuide.md]: image
The `<<sth>>` is used in UML to denote enumeration. Maybe you can reconsider the symbol here?

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,34 +1,614 @@
...
+
+On startup, Duke invokes the loadSavedAppointment() and loadSavedPatient() methods in Storage. This allows the program 
+to retrieve previously stored data from a .txt file and convert it into the static AppointmentList and PatientList objects for use
+within the program. 
+
+The Storage object creates a Scanner object that will parse individual lines in the .txt file, convert them into
+new Appointments, and then add them to an ArrayList called `appointmentListToReturn`. This `appointmentListToReturn` will be passed back to Duke to
+construct the static AppointmentList. The sequence diagram is shown below:
+
+![](images/loadsavedappt_seq.PNG)
```
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md]: image

Maybe `appointmentListToReturn` and `s` should have activation bars for constructors?

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,34 +1,614 @@
...
+
+#### 2.2.4 A&D command module 
+
+The command module consist of 11 different classes, where each class does a different command by itself. 
+These classes allows the patients and appointments to be added into HAMS, allows the updating of patient and 
+appointment details, allows the admin to delete patients and appointment information and allows of listing of the 
+different patients and the appointments in HAMS. In addition, the classes also deals with displaying the list of 
+commands and as well as to allowing HAMS to exit.
+
+![](images/Command%20class%20diagram.png)
```
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md:235]: image

Consider omitting some details (e.g. private variables) to make the diagram less complicated?

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,34 +1,614 @@
...
+creates the connection from the ```Main``` class to the other classes required such as ```Storage``` class,
+```PatientList``` class and the ```Appointment``` class to name a few. 
+
+##### 2.2.4.1 AddPatientClass
+
+To add a patient, the ```AddPatientCommand``` class is used. For this ```AddPatientCommand``` class, it serves as the 
+façade class for the ```Main```, ```Patient``` , ```PatientList``` and the ```Storage``` class to interact with one 
+another. Also, to uniquely identify a patient, an unique patientId number is assigned to each patient when they are first added into the patient list.
+
+![](images/AddPatientDiagram.png)
```
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md:251]: Same problem here: consider omitting some details (e.g. private variables) to make the diagram less complicated?

File: docs/DeveloperGuide.md (hunk 6 above)
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md:278]: image

Should there be a cut at the lower right corner of the `loop` rectangle?

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (JTWang2000) [inline docs/DeveloperGuide.md:235]: If editPaitent and editAppointment is elaborated here, is it okay to not include the class diagram of these two classes. Since there are so many objects here, we may not check things well. It may be a little complicated.

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,34 +1,614 @@
...
+1. This class' object is first created by the ```Parser``` class, where it is then returned to the ```Main``` class to have its
+```execute(Ui ui, Storage storage)``` function be called. 
+
+2. When the ```Main``` class calls the 
+```execute(Ui ui, Storage storage)``` function, ```ListPatientCommand``` will call upon the ```Ui```’s 
+```showEntirePatientList()``` function to display the list of patients.
+
+Below shows the sequence diagram for ```ListPatientCommand``` class
+
+![](images/ListPatientCommandSequence.png)
```
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md:344]: image

Same here: should there be a cut at the lower right corner of the loop rectangle?

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,34 +1,614 @@
...
+class’s ```saveAppoinmentList()``` function to save the updated appointment list. 
+
+5. Upon successfully adding the 
+```Appointment``` object into the appointment list, it will call upon the ```Ui``` class’ 
+```showAppointmentAddSuccess()``` function to display the success of adding the ```appointment``` into the appointment 
+list.
+
+Below shows the sequence diagram for ```AddAppointmentCommand``` class
+
+![](images/AddAppointCommandSequence.png)
```
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md:322]: image

Same here: should there be a cut at the lower right corner of the loop rectangle?

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,34 +1,614 @@
...
+1. This class' object is first created by the ```Parser``` class, where it is then returned to the ```Main``` class to 
+have its ```execute(Ui ui, Storage storage)``` function be called. 
+
+2. When the ```Main``` class calls the 
+```execute(Ui ui, Storage storage)``` function, ```ListAppointmentCommand``` will call upon the ```Ui```’s 
+```showEntireAppointmentList()``` function to display the list of appointments.
+
+Below shows the sequence diagram for ```ListAppointmentCommand``` class
+
+![](images/ListAppointmentCommandSequence.png)
```
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md:362]: image

Same here: should there be a cut at the lower right corner of the loop rectangle?

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,34 +1,614 @@
...
+![](images/SD_parser/Slide2.JPG)
+
+Sequence Diagram for `editp`
+![](images/SD_parser/Slide3.JPG)
+
+Sequence Diagram for `deletep`
+![](images/SD_parser/Slide4.JPG)
+
+Sequence Diagram for `adda`
+![](images/SD_parser/Slide5.JPG)
```
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md:942]: image

Should the return arrow touch the side of the activation bar?

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,34 +1,614 @@
...
+
+Sequence Diagram for `deletea`
+![](images/SD_parser/Slide7.JPG)
+
+Sequence Diagram when it is an unknown command
+![](images/SD_parser/Slide8.JPG)
+
+Sequence Diagram for the creation of the command Object
+
+![](images/SD_parser/Slide9.JPG)
```
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md]: image

Should the starting point of the `getCommandObject` arrow and the ending point of the return arrows align?

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,34 +1,614 @@
...
+5. At the end of the execution, a reference to the command object will be returned. 
+
+>![](images/SD_parser/Capture2.JPG)
+
+>![](images/SD_parser/Capture.JPG)
+
+Sequence Diagram when `parseCommand(userInput)` is initially called
+![](images/SD_parser/Slide1.JPG)
+
+Sequence Diagram for `addp`
```
Comment (Zhilin-Huang) [inline docs/DeveloperGuide.md]: Consider adding more explanations for sequence diagrams in this section?

**Review by JTWang2000 — COMMENTED** [review]
Overall, it is a very good Developer Document. There are proper amount of visuals so that we can understand your design very clearly. However, please note two common erros in sequence diagram: 1. Activation bar too long and 2. Broken activation bar. Also, some UMLs are a little complicated for users.  Overall, I do repect your effort and the great work.

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (yolanmz) [inline docs/DeveloperGuide.md:235]: Should the inheritance here be empty arrow instead?

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,34 +1,614 @@
...
+![](images/AddPatientCommandSequence.png)
+ 
+
+##### 2.2.4.2 AddAppointmentClass
+
+To add an appointment, the ```AddAppointmentCommand``` class is used. For this ```AddAppointmentCommand``` class, it 
+serves as a façade class for the ```Main```, ```Appointment```, ```AppointmentList``` and the ```Storage``` class to 
+interact with one another. 
+
+![](images/AddAppointmentDiagram.png)
```
Comment (yolanmz) [inline docs/DeveloperGuide.md:293]: Consider omitting less important details here to keep the diagram simple?

Comment (yolanmz) [inline docs/DeveloperGuide.md:293]: Also maybe keep the arrows straight and nonoverlapping would make it easier to read?

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (yolanmz) [inline docs/DeveloperGuide.md:251]: Since this subcomponent class is shown very detailedly here, you can probably keep it simple in the main command diagram?

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (alexlim510) [inline docs/DeveloperGuide.md]: image
- Should the class name be in pascal case instead?

image
- Is the method call editp.execute() correct? Maybe you can change it to execute() instead.

image
image
- Should the return arrow be dotted instead?

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (alexlim510) [inline docs/DeveloperGuide.md]: image
- When newDate is returned from TimeConverter, shouldn't the activation bar end at the return arrow? Maybe you can consider splitting the activation bar in TimeConverter into 2 bars. Also, you may consider the same for the activation bar in Appointment.
- Are the activation bars too long? You might want to reduce the length of the activation bars such that it ends at the return arrow.

File: docs/DeveloperGuide.md (hunk 8 above)
Comment (alexlim510) [inline docs/DeveloperGuide.md]: image
- Shouldn't there be activation bars for the constructor for the two classes?

image
- Does the activation bar for the Scanner class look out of place? Maybe you can consider extending the length of the activation bar so that the reference frame lies in between the activation bar. Otherwise, you might want to remove the activation bar.
- See [reference frame](https://nus-cs2113-ay1920s2.github.io/website/book/uml/sequenceDiagrams/referenceFrames/)

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,34 +1,614 @@
...
+On startup, Duke invokes the loadSavedAppointment() and loadSavedPatient() methods in Storage. This allows the program 
+to retrieve previously stored data from a .txt file and convert it into the static AppointmentList and PatientList objects for use
+within the program. 
+
+The Storage object creates a Scanner object that will parse individual lines in the .txt file, convert them into
+new Appointments, and then add them to an ArrayList called `appointmentListToReturn`. This `appointmentListToReturn` will be passed back to Duke to
+construct the static AppointmentList. The sequence diagram is shown below:
+
+![](images/loadsavedappt_seq.PNG)
+![](images/loadsavedappt_ref1.PNG)
```
Comment (alexlim510) [inline docs/DeveloperGuide.md]: image
- Maybe you can remove the return arrows if their not really informational for minimal notation?

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,34 +1,614 @@
...
+
+When the static AppointmentList or PatientList has changes, or the program is exiting, saveAppointmentList() or savePatientList() 
+is invoked respectively. This allows the Storage object to back up existing records to a local .txt file.
+
+The Storage object will create a FileWriter object called `fw`. The command will then iterate through the existing AppointmentList
+and parse each Appointment within, converting it to a string. `fw` then writes this string to the .txt file.
+The sequence diagram is shown below:
+
+![](images/saveapptlist_seq.PNG)
```
Comment (alexlim510) [inline docs/DeveloperGuide.md:222]: image
- Should the text in the square brackets be the terminating condition for the loop? Maybe you can change it to "Until all appointments are stored"

File: docs/DeveloperGuide.md (hunk 6 above)
Comment (alexlim510) [inline docs/DeveloperGuide.md:278]: image
- Should the operation arrows be solid and return arrows be dotted?

**Review by alexlim510 — COMMENTED** [review]
There are some mistakes in the sequence diagrams but still good job! There are many sequence diagrams throughout the document which helps to make the document clearer for the reader. (Y)
